src/wallet/migrate.cpp

```diff
@@ -0,0 +1,771 @@
+    if (inner_meta.encrypt_algo != 0) {
+        throw std::runtime_error("BDB builtin encryption is not supported");
+    }
+
+    // Do a DFS through the BTree, starting at root
+    std::vector<uint32_t> pages{inner_meta.root};
+    while (pages.size() > 0) {
+        uint32_t curr_page = pages.back();
+        if (curr_page > inner_meta.last_page) {
+            std::runtime_error("Page number is greater than subdatabase last page");
+        }
+        pages.pop_back();
+        SeekToPage(db_file, curr_page, page_size);
+        PageHeader header(curr_page, inner_meta.other_endian);
+        db_file >> header;
+        if (header.type != PageType::BTREE_LEAF && header.type != PageType::BTREE_INTERNAL) {
+            throw std::runtime_error("Unexpected inner database page type");
+        }
+        switch (header.type) {
+        case PageType::BTREE_INTERNAL:
+        {
+            InternalPage int_page(header);
+            db_file >> int_page;
+            for (const InternalRecord& rec : int_page.records) {
+                if (rec.m_header.IsDeleted()) continue;
+                pages.push_back(rec.page_num);
+            }
+            break;
+        }
+        case PageType::BTREE_LEAF:
+        {
+            RecordsPage rec_page(header);
+            db_file >> rec_page;
+            bool is_key = true;
+            std::vector<std::byte> key;
+            for (const std::variant<DataRecord, OverflowRecord>& rec : rec_page.records) {
+                std::vector<std::byte> data;
+                if (const DataRecord* drec = std::get_if<DataRecord>(&rec)) {
+                    if (drec->m_header.IsDeleted()) continue;
+                    data = drec->data;
+                } else if (const OverflowRecord* orec = std::get_if<OverflowRecord>(&rec)) {
+                    if (orec->m_header.IsDeleted()) continue;
+                    uint32_t next_page = orec->page_number;
+                    while (next_page != 0) {
+                        SeekToPage(db_file, next_page, page_size);
+                        PageHeader opage_header(next_page, inner_meta.other_endian);
+                        db_file >> opage_header;
+                        if (opage_header.type != PageType::OVERFLOW_DATA) {
+                            throw std::runtime_error("Bad overflow record page type");
+                        }
+                        OverflowPage opage(opage_header);
+                        db_file >> opage;
+                        data.insert(data.end(), opage.data.begin(), opage.data.end());
+                        next_page = opage_header.next_page;
+                    }
+                }
+
+                if (is_key) {
+                    key = data;
+                } else {
+                    m_records.emplace(SerializeData{key.begin(), key.end()}, SerializeData{data.begin(), data.end()});
+                    key.clear();
+                }
+                is_key = !is_key;
+            }
```

Comment:
Could add a check that `is_key` is true at the end. If not, there's an odd number of values, which would be odd?

Reply:
Good point. I've added a check earlier that the number of records is even to avoid this kind of issue.